bot/exts/help_channels/_channel.py

```diff
@@ -25,23 +25,43 @@ def get_category_channels(category: discord.CategoryChannel) -> t.Iterable[disco
             yield channel
 
 
-async def get_idle_time(channel: discord.TextChannel) -> t.Optional[int]:
-    """
-    Return the time elapsed, in seconds, since the last message sent in the `channel`.
-
-    Return None if the channel has no messages.
-    """
-    log.trace(f"Getting the idle time for #{channel} ({channel.id}).")
-
-    msg = await _message.get_last_message(channel)
-    if not msg:
-        log.debug(f"No idle time available; #{channel} ({channel.id}) has no messages.")
-        return None
-
-    idle_time = (datetime.utcnow() - msg.created_at).seconds
-
-    log.trace(f"#{channel} ({channel.id}) has been idle for {idle_time} seconds.")
-    return idle_time
+async def get_closing_time(channel: discord.TextChannel) -> t.Optional[int]:
+    """Return the timestamp at which the given help `channel` should be closed."""
+    log.trace(f"Getting the closing time for #{channel} ({channel.id}).")
+
+    if await _message.is_empty(channel):
+        idle_minutes = constants.HelpChannels.deleted_idle_minutes
+    else:
+        idle_minutes = constants.HelpChannels.idle_minutes_others
+
+    last_message_time = await _caches.last_message_times.get(channel.id)
+    claimant_last_message_time = await _caches.claimant_last_message_times.get(channel.id)
+
+    if not (last_message_time or claimant_last_message_time):
+        # Using the old method if we can't get cached info.
+        msg = await _message.get_last_message(channel)
+        if not msg:
+            log.debug(f"No idle time available; #{channel} ({channel.id}) has no messages.")
+            return datetime.min
+
+        # We want to get the time at which a channel should be closed.
+        closing_time = msg.created_at
+        closing_time += timedelta(minutes=idle_minutes)
+
+        return closing_time
+
+    # We want to get the time at which a channel should be closed.
+    last_message_time = datetime.fromtimestamp(last_message_time)
+    claimant_last_message_time = datetime.fromtimestamp(claimant_last_message_time)
+
+    last_message_time += timedelta(minutes=idle_minutes)
+    claimant_last_message_time += timedelta(minutes=constants.HelpChannels.idle_minutes_claimant)
+
+    # The further away closing time is what we should use.
+    closing_time = max(claimant_last_message_time, last_message_time)
```

Comment:
If `_message.is_empty(channel)` returns `True`, this indicates that there isn't a message in the channel. So both cache's would be empty. This would mean we enter the if statement and return before this code is hit.

Reply:
That is incorrect because the cache doesn't get cleared when the channel goes dormant. `is_empty` doesn't mean the channel has no messages at all; it basically means the channel's latest message is the bot's embed.

Reply:
I added code to clear the caches when the channel is unclaimed in a recent commit, so would that mean it's now correct?
https://github.com/python-discord/bot/pull/1470/files#diff-c31098978b8d838eadb52b2bf23494072cc66686e738d7f99c4f59e6495bfe6fR380-R381

Reply:
Maybe, but is there a way to fix it without relying on the cache being cleared?

Reply:
How do you feel about returning early, within the `if await _message.is_empty(channel):` block?